File: pto-site/tools/make-icons.py
#!/usr/bin/env python3
"""Иконки сайта из assets/img/favicon.svg.

Зачем: iOS не умеет ставить SVG на домашний экран, а старые браузеры и
некоторые агрегаторы ищут favicon.ico. Без них вместо знака компании
показывается пустой квадрат или буква из адреса.

Что делает: рисует знак в браузере и сохраняет
    assets/img/apple-touch-icon.png   180x180 — для iPhone и iPad
    assets/img/favicon.ico            32 и 16 — для старых браузеров

Файлы уже лежат в репозитории. Запускать нужно, только если поменялся знак:
    pip install playwright pillow && playwright install chromium
    python3 tools/make-icons.py
"""
import io
import pathlib
import sys

ROOT = pathlib.Path(__file__).resolve().parent.parent
SVG = ROOT / "assets" / "img" / "favicon.svg"
OUT = ROOT / "assets" / "img"

try:
    from PIL import Image
    from playwright.sync_api import sync_playwright
except ImportError:
    sys.exit("Нужны playwright и pillow: pip install playwright pillow")


def render(size: int) -> Image.Image:
    """Рисует SVG в браузере и возвращает картинку нужного размера."""
    html = (f'<style>html,body{{margin:0}}svg{{display:block;width:{size}px;'
            f'height:{size}px}}</style>' + SVG.read_text(encoding="utf-8"))
    with sync_playwright() as p:
        browser = p.chromium.launch(
            executable_path="/opt/pw-browsers/chromium-1194/chrome-linux/chrome")
        page = browser.new_context(viewport={"width": size, "height": size},
                                   device_scale_factor=1).new_page()
        page.set_content(html)
        page.wait_for_timeout(200)
        shot = page.screenshot(omit_background=False)
        browser.close()
    return Image.open(io.BytesIO(shot)).convert("RGBA")


def main() -> None:
    big = render(180)
    big.save(OUT / "apple-touch-icon.png")
    print(f"  apple-touch-icon.png  180x180  {(OUT / 'apple-touch-icon.png').stat().st_size // 1024} КБ")
    ico = render(64)
    ico.save(OUT / "favicon.ico", sizes=[(32, 32), (16, 16)])
    print(f"  favicon.ico           32 и 16  {(OUT / 'favicon.ico').stat().st_size // 1024} КБ")


if __name__ == "__main__":
    main()

File: norma-site/scripts/make-icons.mjs
// Собирает растровые значки сайта из public/favicon.svg.
//
// Запуск:  node scripts/make-icons.mjs   (или npm run icons)
// Повторять нужно, только если поменялся сам знак.
//
// Зачем растр, когда есть SVG. Браузеры SVG понимают, а вот значок сайта
// в выдаче Яндекса, иконка на домашнем экране телефона и знак организации
// в микроразметке — это растровые картинки. Без них в выдаче на месте
// значка пустой лист.
//
// Углы. У квадратных значков (apple-touch-icon и знак для микроразметки)
// скругления нет нарочно: iOS и поисковики скругляют картинку сами,
// и уже скруглённый файл даёт двойную рамку с прозрачными уголками,
// которые телефон дорисовывает чёрным.

import { chromium } from 'playwright'
import { readFile, writeFile } from 'node:fs/promises'
import { fileURLToPath } from 'node:url'
import { dirname, resolve } from 'node:path'

const here = dirname(fileURLToPath(import.meta.url))
const pub = resolve(here, '../public')

const svg = await readFile(resolve(pub, 'favicon.svg'), 'utf8')
const square = svg.replace(/ rx="\d+"/, '') // тот же знак, но без скруглений

const ICONS = [
  { file: 'favicon-32.png', size: 32, art: svg },
  { file: 'favicon-192.png', size: 192, art: svg },
  { file: 'apple-touch-icon.png', size: 180, art: square },
  { file: 'logo-512.png', size: 512, art: square },
]

const browser = await chromium.launch()
try {
  for (const { file, size, art } of ICONS) {
    const page = await browser.newPage({ viewport: { width: size, height: size }, deviceScaleFactor: 1 })
    await page.setContent(
      `<style>html,body{margin:0;padding:0;background:transparent}svg{display:block;width:${size}px;height:${size}px}</style>${art}`,
    )
    const shot = await page.screenshot({ omitBackground: true, type: 'png' })
    await writeFile(resolve(pub, file), shot)
    await page.close()
    console.log(`  ${file}  ${size}×${size}`)
  }
} finally {
  await browser.close()
}
console.log('✓ Значки собраны из favicon.svg')
